File: arraysAndHashing/maxNumberOfBalloons/maxNumberOfBalloons.js
// Given a string text, you want to use the characters of text to
// form as many instances of the word "balloon" as possible.

// You can use each character in text at most once. Return the
// maximum number of instances that can be formed.

// source:https://leetcode.com/problems/maximum-number-of-balloons/

var maxNumberOfBalloons = function(text) {
    let balloon = {
        "b": 1,
        "a": 1,
        "l": 2,
        "o": 2,
        "n": 1
    };

    let obj = {};
    for (let i = 0; i < text.length; i++) {
        obj[text[i]] = (obj[text[i]] || 0) + 1;
    }
    let freq = [];
    for (let letter in balloon) {
        if (!Math.floor(obj[letter]/balloon[letter])) return 0;
        freq.push(Math.floor(obj[letter]/balloon[letter]))
    }
    return Math.min(...freq)

};

File: arraysAndHashing/maxNumberOfBalloons/max_number_of_balloons.py
# Given a string text, you want to use the characters of text to form as many
# instances of the word "balloon" as possible.

# You can use each character in text at most once. Return the maximum number of
# instances that can be formed.

class Solution:
    def maxNumberOfBalloons(self, text: str) -> int:
        letter_count = {}
        for letter in text:
            if letter in "balloon":
                letter_count[letter] = letter_count.get(letter, 0) + 1
        count = 0

        return min(
            letter_count.get('b', 0),
            letter_count.get('a', 0),
            letter_count.get('l', 0) // 2,
            letter_count.get('o', 0) // 2,
            letter_count.get('n', 0)
        )


# space: O(n)
# time: O(1)
